docs/writing_rules/examples.md

```diff
@@ -0,0 +1,232 @@
+
+## Auditing Dangerous Function Use
+
+Using Semgrep to audit dangerous function calls is easy.
+
+1. Match a function call by name.
+2. Filter out hardcoded strings.
+3. Look explicitly for dangerous keyword arguments.
+
+Let's do an example with the `subprocess` module in Python.
+
+**Match the function call by name.** The ellipsis operator `...` abstracts away whole segments of code. Effectively, it says "I don't care about what's in here."
+
+```yaml
+patterns:
+- pattern: subprocess.call(...)
+```
+
+```python
+import subprocess
+import sys
+
+subprocess.call("echo 'hello'") # Matches here
+subprocess.call("grep -R {} .".format(sys.argv[1])) # Matches here
+subprocess.call("grep -R {} .".format(sys.argv[1]), shell=True) # Matches here
+subprocess.call("grep -R {} .".format(sys.argv[1]), shell=True, cwd="/home/user") # Matches here
+subprocess.run("grep -R {} .".format(sys.argv[1]), shell=True) # Doesn't match here
+```
+
+https://semgrep.live/eq5Z
+
+**Filter out hardcoded strings.** The ellipsis operator can be used inside quotes to represent any string literal. We can filter out static strings by using this with the `pattern-not` clause.
+
+```yaml
+patterns:
+- pattern-not: subprocess.call("...")
+- pattern: subprocess.call(...)
+```
+
+```python
+import subprocess
+import sys
+
+subprocess.call("echo 'hello'") # Doesn't match here anymore!
+subprocess.call("grep -R {} .".format(sys.argv[1])) # Matches here
+subprocess.call("grep -R {} .".format(sys.argv[1]), shell=True) # Matches here
+subprocess.call("grep -R {} .".format(sys.argv[1]), shell=True, cwd="/home/user") # Matches here
+subprocess.run("grep -R {} .".format(sys.argv[1]), shell=True) # Doesn't match here
+```
+
+https://semgrep.live/v8X8
+
+**Look explicitly for dangerous keyword arguments.** You may want to match only when [certain keyword arguments are present](https://docs.python.org/3/library/subprocess.html#security-considerations). We can write keyword arguments just like in Python into our pattern. Combined with the ellipsis operator, this pattern will match if `shell=True` appears at the end of the sequence of arguments.
```

Comment:
Suggested rewording, to make it clear about the impact of this even if the person isn't a security person or familiar with Python:

-------

**Look explicitly for dangerous keyword arguments.** You may want to match only when [certain keyword arguments are present](https://docs.python.org/3/library/subprocess.html#security-considerations). For example, when `subprocess.call` is passed the keyword argument `shell=True`, Python won't auto-escape shell metacharacters that are passed in, which, if an attacker has control over the input, may lead to them being able to run arbitrary shell commands.

We can match keyword arguments just like in Python into our pattern. Combined with the ellipsis operator, this pattern will match if `shell=True` appears at the end of the sequence of arguments.